### PR TITLE
Charge for WASM instantiation

### LINKED ISSUE #1568: Charge for WASM module instantiation
Need to charge the user for creating the WASM module from the code binary.

### PR DESCRIPTION
**Pull request by ark0f**
Resolves #1568

### CODE REVIEW

File: core-backend/common/src/lib.rs [hunk 1]
```diff
@@ -194,12 +194,12 @@ pub trait Environment<E: Ext + IntoExtInfo + 'static>: Sized {
...
     /// 3) Runs `pre_execution_handler` to fill the memory before running instance.
     /// 4) Instantiate external funcs for wasm module.
-    /// 5) Run instance setup starting at `entry_point` - wasm export function name.
+    fn new(ext: E, binary: &[u8], mem_size: WasmPageNumber) -> Result<Self, Self::Error>;
+
+    /// Run instance setup starting at `entry_point` - wasm export function name.
     fn execute<F, T>(
-        ext: E,
-        binary: &[u8],
+        self,
```
Comment (breathx) [inline core-backend/common/src/lib.rs:201]:
```suggestion
        &self,
```

Can't we use self by ref for future purposes?

File: core-backend/sandbox/src/env.rs [hunk 2]
```diff
@@ -173,19 +165,39 @@ where
...
             err: FuncError::Terminated(TerminationReason::Success),
         };
 
-        let mut instance = match Instance::new(binary, &env_builder, &mut runtime) {
+        let instance = match Instance::new(binary, &env_builder, &mut runtime) {
             Ok(inst) => inst,
             Err(e) => return Err((runtime.ext.gas_amount(), ModuleInstantiation(e)).into()),
         };
 
+        Ok(Self { instance, runtime })
```
Comment (breathx) [inline core-backend/sandbox/src/env.rs:181]:
```suggestion
        match Instance::new(binary, &env_builder, &mut runtime) {
            Ok(instance) => Ok(Self { instance, runtime }),
            Err(e) => Err((runtime.ext.gas_amount(), ModuleInstantiation(e)).into()),
        }
```

Probably `map` and `map_err` will also be possible to use despite the compiler strange stuff

File: core-processor/src/processor.rs [hunk 3]
```diff
@@ -127,25 +127,35 @@ pub fn prepare(
...
+            &block_config.allocations_config,
+            &mut gas_counter,
+            &mut gas_allowance_counter,
+            program.get_allocations(),
+            program.static_pages(),
+            dispatch.context().is_none() && matches!(dispatch.kind(), DispatchKind::Init),
+            subsequent_execution,
+        )?;
+        executor::charge_gas_for_instantiation(
+            block_config.module_instantiation,
```
Comment (breathx) [inline core-processor/src/processor.rs:141]: how can you charge some "constant" to block value, if instantiation cost is proportional to code size?

there should be param of code len and cost per byte, than `charge_gas_for_instantiation` calculates it automatically

File: core-processor/src/configs.rs [hunk 4]
```diff
@@ -112,6 +114,8 @@ pub struct BlockConfig {
     pub mailbox_threshold: u64,
     /// Cost for single block waitlist holding.
     pub waitlist_cost: u64,
+    /// Cost for single WASM module instantiation.
+    pub module_instantiation: u64,
```
Comment (breathx) [inline core-processor/src/configs.rs:118]:
```suggestion
    pub module_instantiation_per_byte: u64,
```

File: pallets/gear/src/weights.rs [hunk 5]
```diff
@@ -40,6 +40,7 @@ pub trait WeightInfo {
     fn grow_cost() -> Weight;
```
Comment (breathx) [inline pallets/gear/src/weights.rs:40]: No need to rebench everything.

File: runtime/gear/src/weights/pallet_gear.rs [hunk 6]
```diff
@@ -19,7 +19,7 @@
 //! Autogenerated weights for pallet_gear
```
Comment (breathx) [inline runtime/gear/src/weights/pallet_gear.rs:19]: Same here

File: runtime/vara/src/weights/pallet_gear.rs [hunk 7]
```diff
@@ -19,7 +19,7 @@
 //! Autogenerated weights for pallet_gear
 //!
```
Comment (breathx) [inline runtime/vara/src/weights/pallet_gear.rs:20]: and here

**Review by breathx — CHANGES_REQUESTED** [review]

**Comment by breathx** [comment]
Also code for benchmark should contain different instructions: not the same one repeated

File: core-backend/common/src/lib.rs (hunk 1 above)
Comment (ark0f) [reply core-backend/common/src/lib.rs:205]: Such changes require a lot of refactoring so it must be separate PR

File: core-processor/src/configs.rs [hunk 8]
```diff
@@ -104,6 +108,8 @@ pub struct BlockConfig {
     pub outgoing_limit: u32,
     /// Host function weights.
     pub host_fn_weights: HostFnWeights,
+    /// Static host function weights.
+    pub static_host_fn_weights: StaticHostFnWeights,
```
Comment (grishasobol) [inline core-processor/src/configs.rs:112]: Why static? Why host fn?  Rename it everywhere. Our runtime use sandbox and don't care whether it use host or embedded executor (idiomatically). Actually I don't know what else do you wanna to put inside `static_host_fn_weights`, but can propose you: `executor_backend_weights` or `sandbox_weights`.

**Review by grishasobol — CHANGES_REQUESTED** [review]

**Review by ukint-vs — APPROVED** [review]

File: core-processor/src/configs.rs (hunk 8 above)
Comment (ukint-vs) [reply core-processor/src/configs.rs:113]: Agree

**Review by grishasobol — APPROVED** [review]

File: core-backend/common/src/lib.rs [hunk 9]
```diff
@@ -198,12 +198,12 @@ pub trait Environment<E: Ext + IntoExtInfo<E::Error> + 'static>: Sized {
...
         entries: BTreeSet<DispatchKind>,
```
Comment (breathx) [inline core-backend/common/src/lib.rs:206]: Shouldn't we pass entries in `new` fn and store inside? We don't need to event instantiate when required entry doesn't exist.

File: core-backend/sandbox/src/env.rs [hunk 10]
```diff
@@ -207,7 +217,12 @@ where
...
         };
 
-        let Runtime { err: trap, .. } = runtime;
+        let Runtime {
+            err: trap,
+            ext,
+            memory_wrap,
+            ..
+        } = runtime;
         drop(instance);
```
Comment (breathx) [inline core-backend/sandbox/src/env.rs:226]: why do you drop instance manually?

File: core-processor/src/executor.rs [hunk 11]
```diff
@@ -125,6 +125,27 @@ fn check_memory<'a>(
     Ok(())
 }
 
+pub(crate) fn charge_gas_for_instantiation(
+    weight: u64,
```
Comment (breathx) [inline core-processor/src/executor.rs:129]: I'd rename this into something like `gas_per_byte`

File: core-processor/src/ext.rs [hunk 12]
```diff
@@ -78,6 +78,29 @@ pub struct ProcessorContext {
     pub reserve_for: u32,
 }
 
+impl Default for ProcessorContext {
```
Comment (breathx) [inline core-processor/src/ext.rs:81]: If it used just once in tests, I'd remove this impl

File: core/src/memory.rs [hunk 13]
```diff
@@ -318,6 +318,7 @@ pub trait Memory {
 }
 
 /// Pages allocations context for the running program.
+#[derive(Debug, Default)]
```
Comment (breathx) [inline core/src/memory.rs:321]: I'd remove here default also

File: core/src/message/context.rs [hunk 14]
```diff
@@ -120,7 +120,7 @@ pub struct ContextStore {
 }
 
 /// Context of currently processing incoming message.
-#[derive(Debug, Eq, Hash, Ord, PartialEq, PartialOrd, Decode, Encode, TypeInfo)]
+#[derive(Debug, Default, Eq, Hash, Ord, PartialEq, PartialOrd, Decode, Encode, TypeInfo)]
```
Comment (breathx) [inline core/src/message/context.rs:123]: And here default is meaningless

File: pallets/gear/src/lib.rs [hunk 15]
```diff
@@ -963,7 +966,7 @@ pub mod pallet {
                     GasHandlerOf::<T>::get_origin_key(msg_id)
                         .map_err(|_| b"Internal error: unable to get origin key".to_vec())
                 };
-
+                {}
```
Comment (breathx) [inline pallets/gear/src/lib.rs:969]:
```suggestion
```

File: pallets/gear/src/lib.rs [hunk 16]
```diff
@@ -73,6 +69,12 @@ use sp_std::{
     prelude::*,
 };
 
+#[cfg(feature = "std")]
+type ExecutionEnvironment = gear_backend_wasmi::WasmiEnvironment<Ext>;
+
+#[cfg(not(feature = "std"))]
+type ExecutionEnvironment = gear_backend_sandbox::SandboxEnvironment<Ext>;
```
Comment (breathx) [inline pallets/gear/src/lib.rs:76]: Don't you want to define Ext in the same manner dependent on `lazy-pages` feature to keep everything variable close?

File: pallets/gear/src/manager/mod.rs [hunk 17]
```diff
@@ -79,6 +80,17 @@ pub enum HandleKind {
     Reply(MessageId, ExitCode),
 }
 
+impl fmt::Debug for HandleKind {
```
Comment (breathx) [inline pallets/gear/src/manager/mod.rs:83]: Why?

File: pallets/gear/src/schedule.rs [hunk 18]
```diff
@@ -471,6 +473,18 @@ macro_rules! cost_byte_batched {
     };
 }
 
+impl<T: Config> Default for Schedule<T> {
```
Comment (breathx) [inline pallets/gear/src/schedule.rs:476]: Why?

File: pallets/gear/src/tests.rs [hunk 19]
```diff
@@ -6198,7 +6208,7 @@ mod utils {
                             end $while
                         )
                         (func $init
-                            i32.const 100
+                            i32.const 1024
```
Comment (breathx) [inline pallets/gear/src/tests.rs:6211]: why?

**Review by breathx — APPROVED** [review]

File: core-backend/sandbox/src/env.rs (hunk 10 above)
Comment (breathx) [reply core-backend/sandbox/src/env.rs:226]: I know, it's copy-pasted, but probably it's not actual now

**Pull request merged**